### PR TITLE
Feature/bnl improved code

### PR DESCRIPTION
**Pull request by federiconova**
This contains the improvements from the BNL Hackathon

### CODE REVIEW

File: UserTools/Factory/Factory.cpp [hunk 1]
```diff
@@ -12,9 +12,7 @@ if (tool=="NHits") ret=new NHits;
 if (tool=="test_vertices") ret=new test_vertices;
 if (tool=="WCSimReader") ret=new WCSimReader;
 if (tool=="DataOut") ret=new DataOut;
-#ifdef BONSAIEXISTS
 if (tool=="BONSAI") ret=new BONSAI;
```
[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/Factory/Factory.cpp:16]: done

File: UserTools/nhits/nhits.cpp [hunk 2]
```diff
@@ -76,6 +76,14 @@ bool NHits::Initialise(std::string configfile, DataModel &data){
...
   if(m_stopwatch) Log(m_stopwatch->Result("Initialise"), INFO, m_verbose);
 
+#ifdef GPU
+  //  GPU_daq::nhits_initialize();
+
+  GPU_daq::nhits_initialize_ToolDAQ(ParameterFile,m_data->IDGeom.size(),temp_trigger_search_window, temp_trigger_search_window_step, m_trigger_threshold, temp_trigger_save_window_pre, temp_trigger_save_window_post);
+#endif
+
+
+
```
[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/nhits/nhits.cpp]: ah, good point. fixed.

File: UserTools/Unity.cpp [hunk 3]
```diff
@@ -1,14 +1,11 @@
...
 #include "DummyTool/DummyTool.cpp"
 #include "TriggerOutput/TriggerOutput.cpp"
 #include "WCSimASCIReader/WCSimASCIReader.cpp"
 #include "nhits/nhits.cpp"
 #include "test_vertices/test_vertices.cpp"
 #include "WCSimReader/WCSimReader.cpp"
 #include "DataOut/DataOut.cpp"
-#ifdef BONSAIEXISTS
 #include "BONSAI/BONSAI.cpp"
-#endif //BONSAIEXISTS
```
[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/Unity.cpp:11]: done

File: UserTools/CUDA/library_daq.h [hunk 4]
```diff
@@ -91,8 +108,8 @@ float cerenkov_costheta;
 __constant__ float constant_cerenkov_costheta;
 double twopi;
 bool cylindrical_grid;
-float *device_times_of_flight; // time of flight between a vertex and a pmt
-float *host_times_of_flight;
+unsigned short *device_times_of_flight; // time of flight between a vertex and a pmt
```
[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/CUDA/library_daq.h]: good idea, done

File: UserTools/CUDA/CUDA_Core.cu [hunk 5]
```diff
@@ -19,13 +19,15 @@
...
-__global__ void kernel_correct_times_and_get_n_pmts_per_time_bin(unsigned int *ct){
+__global__ void kernel_correct_times_and_get_n_pmts_per_time_bin(histogram_t *ct){
 
   int time_bin = get_time_bin();
 
   if( time_bin < 0 ) return;
 
+#if defined __HISTOGRAM_UINT__
   atomicAdd(&ct[time_bin],1);
+#endif
```
[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/CUDA/CUDA_Core.cu:30]: uuu, I have no idea why we did that (back in New York). But we never use that function. The only kernel we call here is kernel_correct_times_and_get_histo_per_vertex_shared(histogram_t *ct). In there, there is this statement:
#if defined __HISTOGRAM_UCHAR__
    ct[local_ihit+time_offset] = min(255, ct[local_ihit+time_offset] + temp[local_ihit]);
#else
    ct[local_ihit+time_offset] += temp[local_ihit];
#endif

which I think is ok.

File: configfiles/WCSimReaderTest/ReconDataOutToolConfig [hunk 6]
```diff
@@ -0,0 +1,2 @@
+outfilename FILES/reconstructed.root
+verbose 3
```
[the remark this answers is not in the document]
Comment (federiconova) [reply configfiles/WCSimReaderTest/ReconDataOutToolConfig]: done

File: configfiles/WCSimReaderTest/ReconResetToolConfig [hunk 7]
```diff
@@ -0,0 +1 @@
+verbose 3
```
[the remark this answers is not in the document]
Comment (federiconova) [reply configfiles/WCSimReaderTest/ReconResetToolConfig]: removed

File: UserTools/CUDA/CUDA_Core.cu (hunk 5 above)
[the remark this answers is not in the document]
Comment (tdealtry) [reply UserTools/CUDA/CUDA_Core.cu:30]: Haha ok. Maybe then do something like this?
```suggestion
#if defined __HISTOGRAM_UINT__
  atomicAdd(&ct[time_bin],1);
#else
  printf("This function needs checking before using it without __HISTOGRAM_UINT__ defined"); 
  exit(-1);
#endif
```

[the remark this answers is not in the document]
Comment (federiconova) [reply UserTools/CUDA/CUDA_Core.cu:33]: can't call exit(-1) in a kernel; I just put a return

File: UserTools/Unity.cpp [hunk 8]
```diff
@@ -1,4 +1,3 @@
-#include "../Build.h"
```
Comment (tdealtry) [inline UserTools/Unity.cpp:1]: I think without this `BONSAIEXISTS` will never be defined, so please revert (unless you've tested and I'm wrong)

Comment (federiconova) [reply UserTools/Unity.cpp]: done

**Comment by tdealtry** [comment]
Great thanks. Ready to go @brichards64

**Comment by federiconova** [comment]
Hei Tom or anyone. My email does not work anymore. Can someone post here the zoom link for today's meeting so I can attend?

**Comment by brichards64** [comment]
hey federico, is there another email or slack or something i can send you it on as this is a public webpage so best not to post it here

**Pull request merged**